releng/feature-announce/feature-announce.go

```diff
@@ -0,0 +1,318 @@
+}
+
+func gatherOptions() options {
+	o := options{}
+	flag.IntVar(&o.logLevel, "V", int(log.WarnLevel), "log level from logrus")
+	flag.StringVar(&o.org, "org", "", "The project org")
+	flag.StringVar(&o.repo, "repo", "", "The project repo")
+	flag.BoolVar(&o.dryRun, "dry-run", true, "Should this be a dry run")
+	flag.StringVar(&o.githubTokenFile, "github-token-file", "", "file containing the github token.")
+	flag.StringVar(&o.repositoryPath, "path-to-repository", "", "path to git repository")
+	flag.StringVar(&o.outputFile, "output-file", "", "path to output file, which will be overwritten if it exists")
+	flag.Parse()
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		log.Fatalf("Invalid options: %v", err)
+	}
+
+	log.SetFormatter(&log.JSONFormatter{})
+	log.SetLevel(log.Level(o.logLevel))
+
+	logger := log.StandardLogger().WithField("app", "feature-announce")
+
+	if err := secret.Add(o.githubTokenFile); err != nil {
+		logger.WithError(err).Fatal("error starting secrets agent")
+	}
+
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: string(secret.GetSecret(o.githubTokenFile))},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	githubClient := github.NewClient(tc)
+
+	gitCli := gitCommandLine{
+		directory: o.repositoryPath,
+		logger:    logger,
+	}
+	announcer := featureAnnouncer{
+		githubClient: githubClient,
+		gitCli:       gitCli,
+		options:      o,
+		logger:       logger,
+	}
+
+	err := announcer.generateUpcomingChangesAnnouncement()
+	if err != nil {
+		logger.WithError(err).Fatalf("failed to create announcement")
+	}
+}
+
+type UpcomingChangesAnnouncementData struct {
+	UpcomingChanges []*ReleaseNote
+}
+
+type gitCommandLine struct {
+	directory string
+	logger    *log.Entry
+}
+
+func (g *gitCommandLine) execute(args ...string) (string, error) {
+	args = append([]string{"-C", g.directory}, args...)
+	g.logger.Debugf("executing 'git %v", args)
+	cmd := exec.Command("git", args...)
+	bytes, err := cmd.CombinedOutput()
+	output := string(bytes)
+	if err != nil {
+		return "", fmt.Errorf("git command %q failed: %w (%s)", args, err, output)
+	}
+	return output, nil
+}
+
+type featureAnnouncer struct {
+	options      options
+	gitCli       gitCommandLine
+	githubClient *github.Client
+	logger       *log.Entry
+}
+
+func (f *featureAnnouncer) generateUpcomingChangesAnnouncement() error {
+	latestReleaseTag, err := f.fetchLatestTag()
+	if err != nil {
+		return fmt.Errorf("error fetching latest tag: %w", err)
+	}
+	features, err := f.fetchCodeChanges(latestReleaseTag)
+	if err != nil {
+		return fmt.Errorf("error fetching code changes for tag %s: %w", latestReleaseTag, err)
+	}
+
+	return f.writeUpcomingChangesAnnouncement(features)
+}
+
+func (f *featureAnnouncer) fetchLatestTag() (string, error) {
+	// fetch latest tag from branch
+	//     a)										  b)						 c)
+	// 	  `git tag --list --sort="-version:refname" | grep -vE 'alpha|beta|rc' | head -1`
```

Comment:
When I run this command locally, it returns 'v1.1.1', but we actually want it to return for 'v1.2.0', no?
Will this always be able to determine the approaching Y stream release over a z stream release (which is an issue with the `update-changelog` script in the user-guide repo)?

Reply:
> When I run this command locally, it returns 'v1.1.1', but we actually want it to return for 'v1.2.0', no?

I can't follow. The tag 1.2.0 has not been set yet, has it? So how should it be output of this command?

> Will this always be able to determine the approaching Y stream release over a z stream release (which is an issue with the update-changelog script in the user-guide repo)?

I think the [descending sorting by version](https://git-scm.com/docs/git-tag#Documentation/git-tag.txt---sortltkeygt) should always yield the latest y stream release.

Reply:
Ack, ignore me :/
I was thinking because we should have the v1.2.0-alpha.0 tag, not seeing that alphas are deliberately removed.